src/Connection/Http2Connection.php

```diff
@@ -1467,9 +1478,58 @@ static function () {
 
                             $this->releaseStream($id); // Response has no body, release stream immediately.
                         } else {
-                            $this->trailerDeferreds[$id] = new Deferred;
                             $this->bodyEmitters[$id] = new Emitter;
 
+                            if ($status >= Status::OK && $status < Status::MULTIPLE_CHOICES && $stream->request->getMethod() === 'CONNECT') {
+                                if (($onUpgrade = $stream->request->getUpgradeHandler()) === null) {
+                                    throw new Http2StreamException('CONNECT or upgrade request made without upgrade handler callback', $id, self::CANCEL);
+                                }
+
+                                $stream->bodySizeLimit = \PHP_INT_MAX; // Remove any body size limitation.
```

Comment:
Should we regularly increase it for 32 bit or very-very long running connections?

We probably also have to increase the timeout.

Reply:
The idle watcher should return because there will be a stream in use so long as the tunneled connection in open.